blob/store.go

```diff
@@ -74,27 +104,28 @@ type StoreCloser interface {
 //
 // [1]: https://en.wikipedia.org/wiki/Linearizability
 type KV interface {
-	// Get fetches the contents of a blob from the store. If the key is not
-	// found in the store, Get must report an ErrKeyNotFound error.
-	Get(ctx context.Context, key string) ([]byte, error)
+	KVCore
 
 	// Put writes a blob to the store. If the store already contains the
 	// specified key and opts.Replace is true, the existing value is replaced
 	// without error; otherwise Put must report an ErrKeyExists error.
 	Put(ctx context.Context, opts PutOptions) error
+}
 
-	// Delete atomically removes a blob from the store. If the key is not found
-	// in the store, Delete must report an ErrKeyNotFound error.
-	Delete(ctx context.Context, key string) error
+// CAS represents a mutable set of content-addressed key-value pairs in which
+// each value is identified by a unique, opaque string key.
+type CAS interface {
+	KVCore
 
-	// List calls f with each key in the store in lexicographic order, beginning
-	// with the first key greater than or equal to start.  If f reports an error
-	// listing stops and List returns.  If f reported an ErrStopListing error,
-	// List returns nil; otherwise List returns the error reported by f.
-	List(ctx context.Context, start string, f func(string) error) error
+	// CASPut writes data to a content-addressed blob in the underlying store,
+	// and returns the assigned key. The target key is returned even in case of
+	// error.
+	CASPut(ctx context.Context, data []byte) (string, error)
 
-	// Len reports the number of keys currently in the store.
-	Len(ctx context.Context) (int64, error)
+	// CASKey returns the content address of data without modifying the store.
+	// This must be the same value that would be returned by a successful call
+	// to CASPut on data.
+	CASKey(ctx context.Context, data []byte) (string, error)
```

Comment:
Maybe: remove the error from CASKey? Having it makes usage a bit more awkward, and if we've fixed the hash algorithm there are no further sources of errors.

BUT, it limits future options, if someone comes up with a different implementation that could reasonably error when computing keys.

I dunno, I'm torn, wdyt?

Reply:
Yes, now that we don't need CASKey to be able to transit an RPC boundary, we can remove the error. It was there mainly for the same reason that context is. (I'm inclined to leave the context in-place, since it is already everywhere in this API, but I agree the error could go away)

Reply:
I think it's safe to remove it, and have done so. In our spirit of wanting to move toward the user not having to think about this, I think this is a reasonable choice.